src/Gremlin.Net.CosmosDb/Serialization/TreeJsonConverter.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Gremlin.Net.CosmosDb.Serialization.Trees;
+using Gremlin.Net.CosmosDb.Structure;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Gremlin.Net.CosmosDb.Serialization
+{
+    public class TreeJsonConverter : JsonConverter
+    {
+        public override bool CanWrite => false;
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer) { }
+
+        private static Dictionary<Type, Func<JsonReader, JsonSerializer, object>> _deserializers;
+        public TreeJsonConverter()
+        {
+            _deserializers = _deserializers ?? new Dictionary<Type, Func<JsonReader, JsonSerializer, object>>
+            {
+                [typeof(Tree)] = ReadTree,
+                [typeof(TreeVertexNode)] = ReadTreeVertexNode,
+                [typeof(TreeEdgeNode)] = ReadTreeEdgeNode
+            };
+        }
+
+        public override bool CanConvert(Type objectType)
+        {
+            return _deserializers.TryGetValue(objectType, out var _);
```

Comment:
Using the `Contains()` method here seems more appropriate since you're not using the value at all.